File: sdk/runanywhere-react-native/src/Core/Capabilities/CapabilityProtocols.ts
/**
 * CapabilityProtocols.ts
 *
 * Base protocols and types for capability abstraction.
 * Matches iOS: Core/Capabilities/CapabilityProtocols.swift
 */

// ============================================================================
// Capability Loading State
// ============================================================================

/**
 * Represents the loading state of a capability.
 * Simple state machine: idle -> loading -> loaded OR failed
 */
export type CapabilityLoadingState =
  | { type: 'idle' }
  | { type: 'loading'; resourceId: string }
  | { type: 'loaded'; resourceId: string }
  | { type: 'failed'; error: Error };

/**
 * Create an idle state
 */
export function idleState(): CapabilityLoadingState {
  return { type: 'idle' };
}

/**
 * Create a loading state
 */
export function loadingState(resourceId: string): CapabilityLoadingState {
  return { type: 'loading', resourceId };
}

/**
 * Create a loaded state
 */
export function loadedState(resourceId: string): CapabilityLoadingState {
  return { type: 'loaded', resourceId };
}

/**
 * Create a failed state
 */
export function failedState(error: Error): CapabilityLoadingState {
  return { type: 'failed', error };
}

/**
 * Check if state is idle
 */
export function isIdle(state: CapabilityLoadingState): boolean {
  return state.type === 'idle';
}

/**
 * Check if state is loading
 */
export function isLoading(state: CapabilityLoadingState): boolean {
  return state.type === 'loading';
}

/**
 * Check if state is loaded
 */
export function isLoaded(state: CapabilityLoadingState): boolean {
  return state.type === 'loaded';
}

/**
 * Check if state is failed
 */
export function isFailed(state: CapabilityLoadingState): boolean {
  return state.type === 'failed';
}

/**
 * Get resource ID from state (if available)
 */
export function getResourceId(state: CapabilityLoadingState): string | null {
  if (state.type === 'loading' || state.type === 'loaded') {
    return state.resourceId;
  }
  return null;
}

// ============================================================================
// Capability Operation Result
// ============================================================================

/**
 * Result of a capability operation with timing metadata
 */
export interface CapabilityOperationResult<T> {
  readonly value: T;
  readonly processingTimeMs: number;
  readonly resourceId: string | null;
}

/**
 * Create a capability operation result
 */
export function createOperationResult<T>(
  value: T,
  processingTimeMs: number,
  resourceId?: string | null
): CapabilityOperationResult<T> {
  return {
    value,
    processingTimeMs,
    resourceId: resourceId ?? null,
  };
}

// ============================================================================
// Component Configuration Protocol
// ============================================================================

/**
 * Base interface for component configuration.
 * Matches iOS ComponentConfiguration protocol.
 */
export interface ComponentConfiguration {
  /**
   * Validate the configuration
   * @throws Error if configuration is invalid
   */
  validate?(): void;
}

// ============================================================================
// Capability Protocols
// ============================================================================

/**
 * Base protocol for all capabilities.
 * Matches iOS Capability protocol.
 */
export interface Capability<TConfiguration extends ComponentConfiguration = ComponentConfiguration> {
  /**
   * Configure the capability
   */
  configure(config: TConfiguration): void;

  /**
   * Cleanup resources
   */
  cleanup(): Promise<void>;
}

/**
 * Protocol for capabilities that load models/resources.
 * Matches iOS ModelLoadableCapability protocol.
 */
export interface ModelLoadableCapability<
  TService,
  TConfiguration extends ComponentConfiguration = ComponentConfiguration,
> extends Capability<TConfiguration> {
  /**
   * Whether a model is currently loaded
   */
  readonly isModelLoaded: boolean;

  /**
   * The currently loaded model/resource ID
   */
  readonly currentModelId: string | null;

  /**
   * Load a model by ID
   * @param modelId - The model identifier
   */
  loadModel(modelId: string): Promise<void>;

  /**
   * Unload the currently loaded model
   */
  unload(): Promise<void>;
}

/**
 * Protocol for capabilities that initialize a service without model loading.
 * (e.g., VAD, Speaker Diarization)
 * Matches iOS ServiceBasedCapability protocol.
 */
export interface ServiceBasedCapability<
  TService,
  TConfiguration extends ComponentConfiguration = ComponentConfiguration,
> extends Capability<TConfiguration> {
  /**
   * Whether the capability is ready to use
   */
  readonly isReady: boolean;

  /**
   * Initialize the capability with default configuration
   */
  initialize(): Promise<void>;

  /**
   * Initialize the capability with configuration
   */
  initializeWithConfig(config: TConfiguration): Promise<void>;
}

/**
 * Protocol for capabilities that compose multiple other capabilities.
 * (e.g., VoiceAgent which uses STT, LLM, TTS, VAD)
 * Matches iOS CompositeCapability protocol.
 */
export interface CompositeCapability {
  /**
   * Whether the composite capability is fully initialized
   */
  readonly isReady: boolean;

  /**
   * Clean up all composed resources
   */
  cleanup(): Promise<void>;
}

// ============================================================================
// Capability Metrics Helper
// ============================================================================

/**
 * Helper for tracking capability operation metrics
 */
export class CapabilityMetrics {
  public readonly startTime: number;
  public readonly resourceId: string;

  constructor(resourceId: string) {
    this.startTime = Date.now();
    this.resourceId = resourceId;
  }

  /**
   * Get elapsed time in milliseconds
   */
  get elapsedMs(): number {
    return Date.now() - this.startTime;
  }

  /**
   * Create a result with the current metrics
   */
  result<T>(value: T): CapabilityOperationResult<T> {
    return createOperationResult(value, this.elapsedMs, this.resourceId);
  }
}

// ============================================================================
// Capability Error
// ============================================================================

/**
 * Common errors for capability operations.
 * Matches iOS CapabilityError enum.
 */
export class CapabilityError extends Error {
  constructor(
    message: string,
    public readonly code: CapabilityErrorCode
  ) {
    super(message);
    this.name = 'CapabilityError';
  }

  static notInitialized(capability: string): CapabilityError {
    return new CapabilityError(`${capability} is not initialized`, CapabilityErrorCode.NotInitialized);
  }

  static resourceNotLoaded(resource: string): CapabilityError {
    return new CapabilityError(
      `No ${resource} is loaded. Call load first.`,
      CapabilityErrorCode.ResourceNotLoaded
    );
  }

  static loadFailed(resource: string, underlyingError?: Error): CapabilityError {
    const message = underlyingError
      ? `Failed to load ${resource}: ${underlyingError.message}`
      : `Failed to load ${resource}: Unknown error`;
    return new CapabilityError(message, CapabilityErrorCode.LoadFailed);
  }

  static operationFailed(operation: string, underlyingError?: Error): CapabilityError {
    const message = underlyingError
      ? `${operation} failed: ${underlyingError.message}`
      : `${operation} failed: Unknown error`;
    return new CapabilityError(message, CapabilityErrorCode.OperationFailed);
  }

  static providerNotFound(provider: string): CapabilityError {
    return new CapabilityError(
      `No ${provider} provider registered. Please register a provider first.`,
      CapabilityErrorCode.ProviderNotFound
    );
  }

  static compositeComponentFailed(component: string, underlyingError?: Error): CapabilityError {
    const message = underlyingError
      ? `${component} component failed: ${underlyingError.message}`
      : `${component} component failed: Unknown error`;
    return new CapabilityError(message, CapabilityErrorCode.CompositeComponentFailed);
  }
}

/**
 * Error codes for capability errors
 */
export enum CapabilityErrorCode {
  NotInitialized = 'not_initialized',
  ResourceNotLoaded = 'resource_not_loaded',
  LoadFailed = 'load_failed',
  OperationFailed = 'operation_failed',
  ProviderNotFound = 'provider_not_found',
  CompositeComponentFailed = 'composite_component_failed',
}

File: sdk/runanywhere-flutter/packages/runanywhere/lib/core/capabilities/capability_protocols.dart
//
// capability_protocols.dart
// RunAnywhere SDK
//
// Base protocols and types for capability abstraction.
// Mirrors iOS CapabilityProtocols.swift from RunAnywhere SDK.
//

/// Represents the loading state of a capability.
///
/// Uses Dart sealed classes to mirror iOS enum with associated values.
/// Matches iOS `CapabilityLoadingState` from Core/Capabilities/CapabilityProtocols.swift
sealed class CapabilityLoadingState {
  const CapabilityLoadingState();

  /// Whether this state represents a loaded resource
  bool get isLoaded => this is CapabilityLoaded;

  /// Whether this state represents a loading in progress
  bool get isLoading => this is CapabilityLoading;

  /// Whether this state represents an idle state
  bool get isIdle => this is CapabilityIdle;

  /// Whether this state represents a failed state
  bool get isFailed => this is CapabilityFailed;
}

/// Idle state - no resource loaded
class CapabilityIdle extends CapabilityLoadingState {
  const CapabilityIdle();

  @override
  bool operator ==(Object other) => other is CapabilityIdle;

  @override
  int get hashCode => runtimeType.hashCode;

  @override
  String toString() => 'CapabilityLoadingState.idle';
}

/// Loading state - resource is being loaded
class CapabilityLoading extends CapabilityLoadingState {
  /// The ID of the resource being loaded
  final String resourceId;

  const CapabilityLoading({required this.resourceId});

  @override
  bool operator ==(Object other) =>
      other is CapabilityLoading && other.resourceId == resourceId;

  @override
  int get hashCode => Object.hash(runtimeType, resourceId);

  @override
  String toString() =>
      'CapabilityLoadingState.loading(resourceId: $resourceId)';
}

/// Loaded state - resource is ready to use
class CapabilityLoaded extends CapabilityLoadingState {
  /// The ID of the loaded resource
  final String resourceId;

  const CapabilityLoaded({required this.resourceId});

  @override
  bool operator ==(Object other) =>
      other is CapabilityLoaded && other.resourceId == resourceId;

  @override
  int get hashCode => Object.hash(runtimeType, resourceId);

  @override
  String toString() => 'CapabilityLoadingState.loaded(resourceId: $resourceId)';
}

/// Failed state - resource loading failed
class CapabilityFailed extends CapabilityLoadingState {
  /// The error that caused the failure
  final Object error;

  const CapabilityFailed({required this.error});

  @override
  bool operator ==(Object other) => other is CapabilityFailed;

  @override
  int get hashCode => runtimeType.hashCode;

  @override
  String toString() => 'CapabilityLoadingState.failed(error: $error)';
}

/// Type of resource managed by a capability.
///
/// Matches iOS `CapabilityResourceType` from Core/Capabilities/ManagedLifecycle.swift
enum CapabilityResourceType {
  /// Large language model
  llmModel('llm_model', 'LLM Model'),

  /// Speech-to-text model
  sttModel('stt_model', 'STT Model'),

  /// Text-to-speech voice
  ttsVoice('tts_voice', 'TTS Voice'),

  /// Voice activity detection model
  vadModel('vad_model', 'VAD Model'),

  /// Speaker diarization model
  diarizationModel('diarization_model', 'Diarization Model');

  /// Raw string value for serialization
  final String rawValue;

  /// Human-readable display name
  final String displayName;

  const CapabilityResourceType(this.rawValue, this.displayName);
}

/// Result of a capability operation with timing metadata.
///
/// Matches iOS `CapabilityOperationResult` from Core/Capabilities/CapabilityProtocols.swift
class CapabilityOperationResult<T> {
  /// The operation result value
  final T value;

  /// Processing time in milliseconds
  final double processingTimeMs;

  /// Optional resource ID associated with the operation
  final String? resourceId;

  const CapabilityOperationResult({
    required this.value,
    required this.processingTimeMs,
    this.resourceId,
  });
}

/// Helper for tracking capability operation metrics.
///
/// Matches iOS `CapabilityMetrics` from Core/Capabilities/CapabilityProtocols.swift
class CapabilityMetrics {
  /// When the operation started
  final DateTime startTime;

  /// Resource ID being operated on
  final String resourceId;

  CapabilityMetrics({required this.resourceId}) : startTime = DateTime.now();

  /// Get elapsed time in milliseconds
  double get elapsedMs =>
      DateTime.now().difference(startTime).inMicroseconds / 1000.0;

  /// Create a result with the current metrics
  CapabilityOperationResult<T> result<T>(T value) {
    return CapabilityOperationResult<T>(
      value: value,
      processingTimeMs: elapsedMs,
      resourceId: resourceId,
    );
  }
}

/// Common errors for capability operations.
///
/// Matches iOS `CapabilityError` from Core/Capabilities/CapabilityProtocols.swift
sealed class CapabilityError implements Exception {
  const CapabilityError();

  /// Get human-readable error description
  String get message;

  @override
  String toString() => message;
}

/// Capability is not initialized
class CapabilityNotInitializedError extends CapabilityError {
  final String capability;

  const CapabilityNotInitializedError(this.capability);

  @override
  String get message => '$capability is not initialized';
}

/// Resource is not loaded
class ResourceNotLoadedError extends CapabilityError {
  final String resource;

  const ResourceNotLoadedError(this.resource);

  @override
  String get message => 'No $resource is loaded. Call load first.';
}

/// Resource loading failed
class LoadFailedError extends CapabilityError {
  final String resource;
  final Object? underlyingError;

  const LoadFailedError(this.resource, [this.underlyingError]);

  @override
  String get message =>
      'Failed to load $resource: ${underlyingError ?? "Unknown error"}';
}

/// Operation failed
class OperationFailedError extends CapabilityError {
  final String operation;
  final Object? underlyingError;

  const OperationFailedError(this.operation, [this.underlyingError]);

  @override
  String get message =>
      '$operation failed: ${underlyingError ?? "Unknown error"}';
}

/// Provider not found
class ProviderNotFoundError extends CapabilityError {
  final String provider;

  const ProviderNotFoundError(this.provider);

  @override
  String get message =>
      'No $provider provider registered. Please register a provider first.';
}

/// Composite component failed
class CompositeComponentFailedError extends CapabilityError {
  final String component;
  final Object? underlyingError;

  const CompositeComponentFailedError(this.component, [this.underlyingError]);

  @override
  String get message =>
      '$component component failed: ${underlyingError ?? "Unknown error"}';
}

/// Lifecycle metrics for tracking load/unload operations.
///
/// Matches iOS `ModelLifecycleMetrics` concept.
class ModelLifecycleMetrics {
  final int totalEvents;
  final DateTime startTime;
  final DateTime? lastEventTime;
  final int totalLoads;
  final int successfulLoads;
  final int failedLoads;
  final double averageLoadTimeMs;
  final int totalUnloads;
  final int totalDownloads;
  final int successfulDownloads;
  final int failedDownloads;
  final int totalBytesDownloaded;

  const ModelLifecycleMetrics({
    required this.totalEvents,
    required this.startTime,
    this.lastEventTime,
    this.totalLoads = 0,
    this.successfulLoads = 0,
    this.failedLoads = 0,
    this.averageLoadTimeMs = 0,
    this.totalUnloads = 0,
    this.totalDownloads = 0,
    this.successfulDownloads = 0,
    this.failedDownloads = 0,
    this.totalBytesDownloaded = 0,
  });
}
